### PR TITLE
Adds a source migration plugin that filters out old content

### PR DESCRIPTION
**Pull request by rmarsh1ua**
## Description

New source plugin that filters old content that's being migrated

## Related issues
closes https://github.com/az-digital/az_quickstart/issues/1486

### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
- Minor release changes
   - [X] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_migration/src/Plugin/migrate/source/FilterOldContent.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * a decade. How to easily cut that out of the migration:
+ * @code
+ *  source:
+ *    plugin: filter_old_content
+ *    node_type: uagc_page
+ *    filter_date: "2016-01-01"
+ * @endcode
+ *
+ * @MigrateSource(
+ *   id = "filter_old_content"
```
Comment (joeparsons) [inline modules/custom/az_migration/src/Plugin/migrate/source/FilterOldContent.php:30]: Nitpick: we should prefix the plugin ID with `az_`

Comment (joeparsons) [inline modules/custom/az_migration/src/Plugin/migrate/source/FilterOldContent.php:30]: Nitpick: we should prefix the plugin ID with `az_`

```suggestion
 *   id = "az_filter_old_content"
```

**Comment by danahertzberg** [comment]
To test: Switch plugin of events migration from D7 node to this plugin

**Comment by joeparsons** [comment]
> What do you think @joeparsons @tadean ?

I like the idea of adding this as optional functionality that can be enabled via config to our existing `az_node` source plugin.

**Comment by rmarsh1ua** [comment]
> > What do you think @joeparsons @tadean ?
> 
> I like the idea of adding this as optional functionality that can be enabled via config to our existing `az_node` source plugin. Would you be OK with that approach @rmarsh1ua?

Yea, I think that makes sense. Originally this was meant to sort of extend az_node in a way I thought was kind of boutique but I agree its probably a really common scenario and it makes to incorporate it directly into the more general plugin, az_node.

File: modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php [hunk 2]
```diff
@@ -44,4 +60,28 @@ public function prepareRow(Row $row) {
...
+      else {
+        $date_cutoff = (new \DateTime($this->configuration['filter_date']))->getTimestamp();
+      }
+
+      $query->condition('n.changed', $date_cutoff, '>');
+    }
+
+    return $query;
+  }
+
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php:86]: /**
   * {@inheritdoc}
   */
  public function is_timestamp($timestamp)
{
   return ((string) (int) $timestamp === $timestamp)
   && ($timestamp <= PHP_INT_MAX)
   && ($timestamp >= ~PHP_INT_MAX)
   && (!strtotime($timestamp));
}

File: modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php [hunk 3]
```diff
@@ -44,4 +60,28 @@ public function prepareRow(Row $row) {
...
+   */
+  public function query() {
+    $query = parent::query();
+    /*
+     * Filter date from migration script can be a
+     * string date or a UNIX timestamp so get
+     * it into correct format
+     */
+    if (array_key_exists("filter_date", $this->configuration)) {
+      if (is_int($this->configuration['filter_date'])) {
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php:74]:
```suggestion
      if ($this->is_timestamp($this->configuration['filter_date'])) {
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php (hunk 2 above)
Comment (trackleft) [reply modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php:86]: What do you think about this approach @rmarsh1ua

Comment (rmarsh1ua) [reply modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php:86]: It blows my mind that this isn't something built into PHP as a general validation function. Looks good, but should it be a private function?

Comment (trackleft) [reply modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php:86]: Yeah, I went down a rabbit hole looking for a built-in too, ha. Probably should be private

File: modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php [hunk 4]
```diff
@@ -44,4 +60,55 @@ public function prepareRow(Row $row) {
...
+     */
+    if (array_key_exists("filter_date", $this->configuration)) {
+      if ($this->isTimestamp($this->configuration['filter_date'])) {
+        $date_cutoff = $this->configuration['filter_date'];
+      }
+      else {
+        $date_cutoff = (new \DateTime($this->configuration['filter_date']))->getTimestamp();
+      }
+
+      $query->condition('n.changed', $date_cutoff, '>');
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZNode.php:81]: Possibly should be `>=`

**Review by trackleft — APPROVED** [review]

**Pull request merged**